src/Plutarch/Extra/ExtendedAssetClass.hs

```diff
@@ -0,0 +1,489 @@
+instance
+  (k ~ A_Getter, a ~ CurrencySymbol, b ~ CurrencySymbol) =>
+  LabelOptic "symbol" k AnyTokenAssetClass AnyTokenAssetClass a b
+  where
+  labelOptic = to coerce
+
+{- | We can always \'pull out\' an 'AssetClass' by \'filling in\' an empty
+ 'TokenName'.
+
+ @since 3.14.2
+-}
+instance
+  (k ~ A_Getter, a ~ AssetClass, b ~ AssetClass) =>
+  LabelOptic "assetClass" k AnyTokenAssetClass AnyTokenAssetClass a b
+  where
+  labelOptic = to $ \(AnyTokenAssetClass cs) -> AssetClass cs ""
+
+-- | @since 3.14.2
+instance PConstantDecl AnyTokenAssetClass where
+  type PConstanted AnyTokenAssetClass = PAnyTokenAssetClass
+  type PConstantRepr AnyTokenAssetClass = Data
+  pconstantFromRepr = fromData
+  pconstantToRepr = toData
+
+-- | @since 3.14.2
+instance FromData AnyTokenAssetClass where
+  fromBuiltinData dat =
+    AnyTokenAssetClass <$> case builtinDataToData dat of
+      Constr 0 [dat'] -> fromData dat'
+      _ -> Nothing
+
+-- | @since 3.14.2
+instance UnsafeFromData AnyTokenAssetClass where
+  unsafeFromBuiltinData dat = case builtinDataToData dat of
+    Constr 0 [dat'] -> case fromData dat' of
+      Just cs -> AnyTokenAssetClass cs
+      Nothing ->
+        error "unsafeFromBuiltinData: Did not get a CurrencySymbol for AnyTokenAssetClass"
+    _ -> error "unsafeFromBuiltinData: Could not make AnyTokenAssetClass"
+
+-- | @since 3.14.2
+instance ToData AnyTokenAssetClass where
+  toBuiltinData (AnyTokenAssetClass cs) =
+    dataToBuiltinData $ Constr 0 [toData cs]
+
+{- | Plutarch equivalent to 'AnyTokenAssetClass'.
+
+ @since 3.14.2
+-}
+newtype PAnyTokenAssetClass (s :: S)
+  = PAnyTokenAssetClass (Term s PCurrencySymbol)
+  deriving stock
+    ( -- | @since 3.14.2
+      Generic
+    )
+  deriving anyclass
+    ( -- | @since 3.14.2
+      PEq
+    , -- | @since 3.14.2
+      PlutusType
+    , -- | @since 3.14.2
+      PShow
+    , -- | @since 3.14.2
+      PIsData
+    )
+
+-- | @since 3.14.2
+instance DerivePlutusType PAnyTokenAssetClass where
+  type DPTStrat _ = PlutusTypeNewtype
+
+-- | @since 3.14.2
+instance PUnsafeLiftDecl PAnyTokenAssetClass where
+  type PLifted PAnyTokenAssetClass = AnyTokenAssetClass
+
+{- | An 'AssetClass' whose 'TokenName' is significant somehow.
+
+ @since 3.14.2
+-}
+newtype FixedTokenAssetClass = FixedTokenAssetClass AssetClass
+  deriving stock
+    ( -- | @since 3.14.2
+      Generic
+    , -- | @since 3.14.2
+      Show
+    )
+  deriving
+    ( -- | @since 3.14.2
+      Eq
+    , -- | @since 3.14.2
+      Ord
+    )
+    via AssetClass
+
+-- | @since 3.14.2
+instance PConstantDecl FixedTokenAssetClass where
+  type PConstantRepr FixedTokenAssetClass = Data
+  type PConstanted FixedTokenAssetClass = PFixedTokenAssetClass
+  pconstantFromRepr = fromData
+  pconstantToRepr = toData
+
+-- | @since 3.14.2
+instance FromData FixedTokenAssetClass where
+  fromBuiltinData dat =
+    FixedTokenAssetClass <$> case builtinDataToData dat of
+      Constr 0 [dat'] -> fromData dat'
+      _ -> Nothing
```

Comment:
The encoding here could be a little neater, currently for ExtendedAssetClass, this would be:

```
Constr 0 [Contstr 0 [B currencySymbol ]]
Constr 1 [Constr 0 [B currencySymbol, B tokenName]]
```

Ideally it should be:

```
Constr 0 [B currencySymbol]
Constr 1 [B currencySymbol, B tokenName]
```

This is very minor though. A related nitpick - I'm not sure why the separate newtypes for the different cases are required.

Reply:
I figured we might want this for cases where we know for sure we have the irrelevant case. Happy to get rid of them if you don't need them.